### PR TITLE
fix(eventhubs): honor transient retry classification

### PR DESCRIPTION
**Pull request by sagar0207**
## Summary

Improve the Event Hubs producer retry loop so it uses the transient classification already carried
by `EventHubsException` and allows callers to cancel retry backoff through
`Azure::Core::Context`.

Azure DevOps work item:
[39045656](https://msazure.visualstudio.com/One/_workitems/edit/39045656)

Related GitHub workstream: #7254

## Changes

- Retry `EventHubsException` only when `IsTransient` is `true`.
- Treat empty AMQP error conditions as transient.
- Retry only an explicit allowlist of known transient AMQP conditions.
- Stop arbitrary unknown and known non-transient AMQP conditions, including authorization,
  precondition, message-size, quota, ownership, lock-loss, redirect, and cancellation failures.
- Preserve bounded `std::runtime_error` retries for reachable AMQP send failures on both backends.
- Stop immediately for:
  - Authorization failures.
  - Precondition failures.
  - Message-size failures.
  - Quota/resource-limit failures.
  - Producer/consumer ownership failures.
  - Authentication failures.
  - Cancellation.
- Pass the send operation context into `RetryOperation`.
- Preserve a successful result if cancellation races after the operation has already completed;
  failed results still surface cancellation before retry.
- Replace unconditional backoff sleeps with context-aware waits that check cancellation every
  100 ms.
- Treat `MaxRetries` as retries after the initial attempt:
  - `MaxRetries == 0` still performs the initial attempt.
  - The first backoff uses attempt 1.
  - No backoff occurs after the final failed attempt.
- Copy retry options instead of moving them out of `ProducerClientOptions`.
- Add focused regression coverage for empty/unknown classification, terminal conditions, AMQP
  runtime errors, retry accounting, final exception preservation, and context cancellation.
- Add a Core AMQP unit test that deterministically covers custom `AmqpErrorCondition` construction
  after the coverage counter reset.

## Validation

Configured and built from the repository root with:

```text
CMAKE_BUILD_TYPE=Debug
BUILD_TESTING=ON
DISABLE_RUST_IN_BUILD=ON
DISABLE_AZURE_CORE_OPENTELEMETRY=ON
Generator=Ninja
```

CMake confirmed:

```text
Disabling Rust based functionality in build.
Using vendored uamqp
Using uAMQP.
```

Results:

- `azure-messaging-eventhubs-test` built successfully with MSVC warnings-as-errors.
- Focused retry suite: **22 passed, 0 failed**.
- Complete Event Hubs playback suite: **55 passed, 34 expected live-only skips, 0 failed**.
- Core AMQP value-model suite: **27 passed, 0 failed**.
- Repository `clang-format` applied.
- `git diff --check` passed.
- Final read-only code review reported no significant issues.

## Scope

This PR does not:

- Rebuild producer links after detach.
- Refresh CBS tokens.
- Recover receivers or resume from the last offset.
- Change non-string `x-opt-offset` handling.
- Add per-attempt `TryTimeout`.
- Add the extra server-busy delay.
- Redesign Rust AMQP error classification.

### CODE REVIEW

File: sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp [hunk 1]
```diff
@@ -6,94 +6,115 @@
...
 
-      if (ShouldRetry(result, retryCount, retryAfter))
-      {
-        lastException = nullptr;
-        retryCount++;
-        std::this_thread::sleep_for(retryAfter);
-      }
-      else
+      if (!ShouldRetry(result, retryCount, retryAfter))
       {
```
[the remark this answers is not in the document]
Comment (sagar0207) [reply sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp:57]: Fixed in a3f36fa4. The post-operation cancellation check now runs only when the operation returns false, so a successful send remains successful even if cancellation races immediately afterward. Failed results still surface cancellation before retry. Added SuccessfulOperationWinsCancellationRace and FailedOperationSurfacesCancellationBeforeRetry; the focused suite passes 22/22 and the full playback suite passes 55 tests with 34 expected live-only skips.

File: sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp [hunk 2]
```diff
@@ -105,25 +129,48 @@ bool Azure::Messaging::EventHubs::_detail::RetryOperation::ShouldRetry(
...
   }
 
-  retryAfter = CalculateExponentialDelay(attempt, jitterFactor);
+  retryAfter = CalculateExponentialDelay(attempt + 1, jitterFactor);
+  Log::Write(Logger::Level::Informational, std::string("Operation failed and will be retried."));
 
   return true;
 }
 
+bool Azure::Messaging::EventHubs::_detail::RetryOperation::ShouldRetry(
```
Comment (j7nw4r) [inline sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp:127]: `IsTransient` now decides the whole retry, and it has one writer: `IsErrorTransient` in `eventhubs_utilities.cpp:128`. That function allowlists eight conditions and returns false for everything else, an empty condition included. An `AmqpError` with no condition gets one attempt.

Empty errors are routine on uAMQP. A released or modified disposition carries no error, so the parser leaves the `AmqpError` empty (`impl/uamqp/amqp/message_sender.cpp:576-589`). AMQP released means the broker did not process the message, so a resend is safe, and that is a retry you are now losing. Detach produces the same empty error by another route: `link.c:655` settles every pending delivery with a NULL state before it fires the detach subscription at `link.c:657`, so the send callback copies `m_savedMessageError` while it is still empty (`message_sender.cpp:563-565`). Retrying a detach would not have helped anyway, because nothing rebuilds the sender. The caller still ends up with an `EventHubsException` carrying an empty `ErrorCondition` and an empty `what()`, which gives you nothing to log.

Flip the default. Make `IsErrorTransient` return true unless the condition sits on an explicit terminal list: `UnauthorizedAccess`, `LinkPayloadSizeExceeded`, `PreconditionFailed`, `ResourceDeleted`, `NotAllowed`, `NotImplemented`, `DecodeError`, `InvalidField`, `FrameSizeTooSmall`, `ArgumentError`, `ArgumentOutOfRangeError`, `EntityDisabledError`, `PublisherRevokedError`, `EntityAlreadyExists`, `OperationCancelled`. This overload stays as it is. You also pick up `ResourceLimitExceeded`, which is quota throttling and should retry.

Two follow-ons. `IsTransient` is public, so this is a public behavior change and wants a changelog line. Playback reaches neither path, so please add a unit test that pushes an empty-condition `AmqpError` through the loop.

File: sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp [hunk 3]
```diff
@@ -6,94 +6,118 @@
...
+#include <system_error>
 #include <thread>
 
 namespace {
-// The set of AMQP error conditions that should be treated as fatal conditions.
-constexpr const char* AmqpFatalConditions[] = {"amqp:link:message-size-exceeded"};
+constexpr std::chrono::milliseconds CancellationCheckInterval{100};
 
-bool IsFatalException(Azure::Messaging::EventHubs::EventHubsException const& ex)
+bool IsRetryableSystemError(std::error_code const& errorCode)
```
Comment (j7nw4r) [inline sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp]: This branch cannot fire. I grepped all of `sdk/`, vendored azure-uamqp-c included, and `system_error` appears only in this file and its new test.

Inside the retry lambda the reachable throws are `std::out_of_range` from `m_senders.at()` and `std::runtime_error` from `QueueSendInternal`. Everything else on the path is C, and uAMQP and azure-c-shared-utility report socket and TLS failures through return codes. That leaves the `std::recursive_mutex` behind `Connection::Lock` (`connection_impl.hpp:135`) and the `std::condition_variable` in `WaitForResult`. Their failure codes sit outside your list, with one theoretical exception: `recursive_mutex::lock` raises `resource_unavailable_try_again` if the recursion count overflows, and the nesting here is bounded.

So `IsRetryableSystemError` never returns true in practice, and the catch only logs and rethrows. I would delete it along with the catch clause, the two tests behind them, and the allowlist sentence in the changelog. Name a caller that does throw `std::system_error` and I will drop this.

File: sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp [hunk 4]
```diff
@@ -6,94 +6,118 @@
...
+        Log::Write(Logger::Level::Warning, std::string("System error while trying: ") + e.what());
+      }
+      if (!IsRetryableSystemError(e.code()) || !ShouldRetry(false, retryCount, retryAfter))
+      {
+        throw;
       }
-      // We assume that all exceptions other than EventHubs exceptions might be retriable.
-      if (ShouldRetry(false, retryCount, retryAfter))
+    }
+#if ENABLE_RUST_AMQP
```
Comment (j7nw4r) [inline sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp]: The `std::runtime_error` catch sits behind `#if ENABLE_RUST_AMQP`, so on uAMQP this drops a retry that `main` had. `main` caught `std::exception` and retried it (`retry_operation.cpp:73`). The replacement is the `std::system_error` branch, which never matches on this path.

The uAMQP send path throws exactly one `std::runtime_error`, and it is not rare. `QueueSendInternal` throws "Could not send message" whenever `messagesender_send_async` returns NULL (`impl/uamqp/amqp/message_sender.cpp:538`). That happens with the sender in ERROR state (`vendor/azure-uamqp-c/src/message_sender.c:1063`). The same NULL comes back on a healthy OPEN link when `send_one_message` fails (`message_sender.c:1129-1136`), which is a transport write failure, and a resend is the right answer to it.

Please move the catch out from behind the guard so uAMQP retries it too. Retrying is safe here. An ERROR state sender returns NULL again, so the loop rethrows after the last attempt, which is what `main` did. A detached sender takes a different route, because `send_async` queues the message instead of returning NULL.

One thing holds either way. The caller currently receives a bare `std::runtime_error` reading "Could not send message" with no AMQP detail on it. If you keep the fail fast, please wrap it in an `EventHubsException` so callers keep a typed error, and put the change in the description.

**Review by j7nw4r — CHANGES_REQUESTED** [review]
Three findings, all on the retry classification. The first one is the blocker: the new gate refuses to retry an AMQP error with an empty condition, and empty conditions arrive on two ordinary uAMQP paths. The other two are a dead allowlist and a retry that the uAMQP build silently lost. Details are inline.

File: sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp (hunk 4 above)
Comment (sagar0207) [reply sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp]: Addressed in 2de76381. The bounded std::runtime_error catch is now backend-independent, with OperationCancelledException still rethrown first. This preserves main behavior for uAMQP messagesender_send_async failures. Added RethrowsLastRuntimeErrorWhenRetriesExhausted using the reachable Could not send message shape.

File: sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp (hunk 2 above)
Comment (sagar0207) [reply sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp:127]: Addressed in 2de76381. Empty AMQP conditions are explicitly transient so uAMQP undelivered sends keep their retry. I kept a strict allowlist rather than defaulting arbitrary future conditions to transient: Timeout, ServerBusy, InternalError, LinkDetachForced, ConnectionForced, ConnectionFramingError, ProtonIo, NotFound, and IllegalState retry; unknown and all other known conditions stop. ResourceLimitExceeded and LinkStolen remain terminal, and redirects remain terminal because this client discards redirect metadata and retries the same cached sender. Added tests for empty, every allowlisted condition, arbitrary unknown, and all remaining known conditions; changelog updated.

File: sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp (hunk 3 above)
Comment (sagar0207) [reply sdk/eventhubs/azure-messaging-eventhubs/src/retry_operation.cpp]: Addressed in 2de76381. Removed IsRetryableSystemError, the std::system_error catch and include, both synthetic system-error tests, and the system-error changelog wording. Repository search found no production caller for those network codes.

**Review by j7nw4r — APPROVED** [review]

**Review by j7nw4r — APPROVED** [review]

**Review by LarryOsterman — APPROVED** [review]
